### PR TITLE
Post: Clear Separation of Specification and Implementation in Dafny

### PR DESCRIPTION
**Pull request by atomb**

### CODE REVIEW

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 1]
```diff
@@ -0,0 +1,610 @@
...
+proving properties of the specification, and proving equivalence (or
+refinement) between the specification and the implementation.
+
+To get started, let's go over the tools Dafny provides. The two most
+common computational descriptions in Dafny are terminating,
+deterministic functions and imperative, stateful methods.
+
+By writing a `function`, you can describe a mathematical object that
+maps each input deterministically to a single output. Functions are
+typically concise, and, given their genesis as one of the
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:24]: I don't think it is because of their genesis that they are easy to reason about. It is because of referential transparency.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 2]
```diff
@@ -0,0 +1,610 @@
...
+foundational concepts in mathematics, they tend to be the most
+straightforward to reason about. Because of this, the bodies of
+functions are made directly available to the verifier by default,
+unless a function is marked `{:opaque}`. Functions can often also be
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:28]: Maybe more specific than "often". Something like, if they are computable, ...

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 3]
```diff
@@ -0,0 +1,610 @@
...
+compiled to executable programs, though sometimes inefficiently (and
+not at all if they make use of infinite data types).
+
+By writing a `method`, you can describe an imperative computation that
+performs a sequence of steps to yield an output state or a set of output
+states from each initial program state. Methods more closely match the
+operations performed by actual computing hardware, and are capable of
+describing non-deterministic or non-terminating computations. They are
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:36]: One subtllty to keep in mind is var x :| P(x); x which is an expression and is at least not referentially transparent, and I don't know if it qualifies as non-deterministic.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+    reveals StackModel, PushModel, PopModel, EmptyModel, IsEmptyModel
+    provides PushPopVal, PushPopStack
+
+  // We provide a concrete type to model a stack for simplicity of
+  // proof, since Dafny knows how to prove things about sequences. Note,
+  // however, that most clients don't get to see what this type is.
+  type StackModel<T> = seq<T>
+
+  // Check to see whether a stack is empty.
+  ghost function IsEmptyModel<T>(stk: StackModel<T>) : bool
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:112]: Why ghost?

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+Several things are important about this implementation. First, it
+provides a number of functions with bodies that describe one way of
+constructing a stack out of built-in primitives. Second, it hides
+the bodies of those functions from most clients, except
+implementation modules that need to prove equivalence with them.
+And, third, it describes the _relationship_ between those operations
+in a couple of lemmas.
+
+One further note: you'll see that some of the operations on a stack
+have pre- and post-conditions, specified on the functions themselves
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:161]: Maybe use the "intrinsic" vs "extrinsic" terminology? Maybe a good opportunity to say a few words about being mindful when choosing one vs the other?

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+rather than in lemmas. These are exactly the properties required to
+ensure that each operation is well-defined without specifying
+anything else about behavior. Besides well-definedness, all
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:164]: I like that

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 7]
```diff
@@ -0,0 +1,610 @@
...
+do. It is possible to constrain type parameters to those that do
+have default values by, for instance, using the declaration `class
+Stack<T(0)>`, to insist that `T` have type characteristic `0`, the
+notation for having a default value.
+
+However, doing this would make the array-based stack have different
+constraints on its type parameter than other stack implementations,
+which isn't ideal. An alternative is to make use of the fact that Dafny
+can define datatypes with ghost alternatives, and that datatypes with
+only one non-ghost alternative with only one argument have no runtime
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:186]: Very nice

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 8]
```diff
@@ -0,0 +1,610 @@
...
+  }
+
+  // This class, that implements the actual linked data structure, is
+  // where the complexity lives. If you already had a linked list
+  // implementation, you could potentially use that, and share the
+  // burden of implementation and verification with any other uses of
+  // linked lists. Although it would require some cleverness to
+  // integrate the ghost model and node validity notions with a
+  // general-purpose list.
+  class Node<T> {
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:402]: Not fond of this design. If you were to add a size operation, the validity predicate would become more complicated. Instead, I would use a root.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 9]
```diff
@@ -0,0 +1,610 @@
...
+
+    // Each node has a value.
+    var val : T
+
+    // Each node points to the next node, which may be null if this is
+    // the last node (the bottom of the stack).
+    var next : Node?<T>
+
+    // The model is the same as in the array-based case.
+    ghost var model : StackModel<T>
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:412]: Maybe a comment about the model being a simpler but perhaps slower implementation, but no impact on performance because of ghost?

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 10]
```diff
@@ -0,0 +1,610 @@
...
+  ghost function NodeModel<T>(node: Node?<T>) : StackModel<T>
+    reads node
+  {
+    if node == null then EmptyModel() else node.model
+  }
+}
+{% endhighlight %}
+
+Finally, given Dafny's built-in types, it's possible to create an
+imperative stack building on the immutable `seq` type. This
```
Comment (jtristan) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:465]: Not sure what you mean, seq is persistent.

**Review by jtristan — APPROVED** [review]

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 1 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:24]: Maybe I should reword it, but what I mean to say is that mathematicians care about being able to reason about things easily, and that's why they've defined functions the way they have.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 4 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:112]: It's used on `StackModel` values rather than stack implementations, and those are always ghost (at least in this example).

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 5 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:161]: Yeah, I like the idea of using that terminology.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 8 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:402]: I derived this one from an example in the test suite. :) Although I think you may be right that another design would be better, I'm not sure how much difference that would make to the overall point this post is attempting to make.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 9 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:412]: Good idea.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 10 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:465]: Yeah, good point. I mean that it's being used in a method, and used in such a way that it can be updated (even if that involves copying) and appears imperative from the interface of the class. But I think I can reword it to be clearer.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 2 above)
Comment (MikaelMayer) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:28]: Opaque keyword, not attribute please :-)

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 11]
```diff
@@ -0,0 +1,610 @@
...
+performs a sequence of steps to yield an output state or a set of output
+states from each initial program state. Methods more closely match the
+operations performed by actual computing hardware, and are capable of
+describing non-deterministic or non-terminating computations. They are
+also generally meant to be compiled to executable programs. However,
+reasoning about methods is somewhat different than reasoning about
+functions. You can prove that a method agrees with a single contract,
+written as `requires`, `ensures`, `modifies`, and `decreases` clauses,
+but you generally do not expand the body of a method as part of
+performing a separate proof (though it is sometimes possible to do so,
```
Comment (MikaelMayer) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:42]: Remove parentheses to increase legibility. Blog posts can also use footnotes if necessary.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 12]
```diff
@@ -0,0 +1,610 @@
...
+Dafny allows you to describe computations in several ways,
+primarily pure functions and imperative methods, each with its
+own tradeoffs. Because of these tradeoffs, it can be useful to
+combine mechanisms and relate them to each other. This post
+describes one particularly effective technique for doing so, by
+clearly separating a _specification_ from an _implementation_,
+proving properties of the specification, and proving equivalence (or
+refinement) between the specification and the implementation.
+
+To get started, let's go over the tools Dafny provides. The two most
```
Comment (MikaelMayer) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:18]: To get started, it would help if you could state an *example* quickly enough of what the problem is about.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 13]
```diff
@@ -0,0 +1,610 @@
...
+describing non-deterministic or non-terminating computations. They are
+also generally meant to be compiled to executable programs. However,
+reasoning about methods is somewhat different than reasoning about
+functions. You can prove that a method agrees with a single contract,
+written as `requires`, `ensures`, `modifies`, and `decreases` clauses,
+but you generally do not expand the body of a method as part of
+performing a separate proof (though it is sometimes possible to do so,
+as will be described in a future post).
+
+These differing attributes come into play particularly starkly when
```
Comment (MikaelMayer) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:45]: Which attributes? Are you referring to clauses?

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 14]
```diff
@@ -0,0 +1,610 @@
...
+given application, but we can still look at how to straightforwardly
+show that multiple implementations, using different techniques, all
+fit the same notion of what it means to be a stack.
+
+To start with, let's specify what it means to be a stack in the
+following module.
+
+{% highlight dafny %}
+module StackSpecification {
+  // Most clients, including clients of implementations, get no details.
```
Comment (MikaelMayer) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:90]: Please try to put this comments after the Dafny code.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 15]
```diff
@@ -0,0 +1,610 @@
...
+`GhostOption` type always has a default value.
+
+{% highlight dafny %}
+module ArrayStackImplementation {
+  // Get access to the implementation details.
+  import opened StackSpecification`Private
+
+  // This allows `elts` below to include default values in unused
+  // positions.
+  datatype GhostOption<T> = ghost None | Some(value: T)
```
Comment (MikaelMayer) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:199]: Wow that's clever! But practically speaking, we still store null? Can you explain?

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 16]
```diff
@@ -0,0 +1,610 @@
...
+    predicate IsEmpty()
+      reads this, top, footprint
+      requires ValidModel()
+      ensures IsEmpty() <==> IsEmptyModel(model)
+    {
+      top == null
+    }
+  }
+
+  // This class, that implements the actual linked data structure, is
```
Comment (MikaelMayer) [inline _posts/2023-08-14-clear-specification-and-implementation.markdown:395]: Same here, please move this in regular text.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 3 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:36]: I think I might gloss over this subtlety for now. It's somewhat beside the overall point of the post.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 14 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:90]: I'm not sure I agree. I find it more readable before the code, and I think it's a more common approach, and therefore less surprising to readers.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 13 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:45]: Reworded.

File: _posts/2023-08-14-clear-specification-and-implementation.markdown (hunk 15 above)
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:199]: I believe the compiler can leave the storage uninitialized, because Dafny can prove that it's never accessed (because that would require using a ghost constructor at runtime).

File: _posts/2023-08-14-clear-specification-and-implementation.markdown [hunk 17]
```diff
@@ -6,45 +6,96 @@ author: Aaron Tomb
...
+    account.Deposit(amount);
+    account.Deposit(-amount);
+    assert account.balance == old(account.balance);
+}
+{% endhighlight %}
+
+However, this is a method rather than a lemma (because lemmas aren't
+allowed to call methods), and even if it were a lemma, it would be
+impossible to use the proven fact anywhere else to prove some
+higher-level property of a larger system. You might argue that you don't
```
[the remark this answers is not in the document]
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:54]: Great thought. I'll add some text about that.

[the remark this answers is not in the document]
Comment (atomb) [reply _posts/2023-08-14-clear-specification-and-implementation.markdown:54]: I might use your example directly. I like it.

**Review by jtristan — APPROVED** [review]

**Review by MikaelMayer — APPROVED** [review]
This is great, thanks. Now you might just want to modify the publication date.

**Comment by CodeNinja89** [comment]
Hi, I tried following your blog and implemented the linked stack example. There seems to be a problem in the Pop() method. I get 2 errors:

1. This postcondition might not hold: ValidModel(). Could not prove model == NodeModel(top)
2. This postcondition might not hold: res == PopModel(old(model)).0

Please find my implementation [here](https://github.com/CodeNinja89/dafnyPractice).

I hope its not some dumb typo on my side. Or maybe the version I use (3.2.2) is different from the one used in the blog.

**Comment by atomb** [comment]
> Hi, I tried following your blog and implemented the linked stack example. There seems to be a problem in the Pop() method. I get 2 errors:
> 
>     1. This postcondition might not hold: ValidModel(). Could not prove model == NodeModel(top)
> 
>     2. This postcondition might not hold: res == PopModel(old(model)).0
> 
> 
> Please find my implementation [here](https://github.com/CodeNinja89/dafnyPractice).
> 
> I hope its not some dumb typo on my side. Or maybe the version I use (3.2.2) is different from the one used in the blog.

Interesting. I've tried re-verifying the code in the blog, and it was successful. But I also can confirm a verification error on the code you linked to. The version of Dafny you're using is pretty old, but I can confirm that the error shows up with 4.4.0, the latest version (and the one I used to verify the code in the blog in the first place).

So, there must be some difference between the two, but I haven't been able to identify exactly what it is yet.
